common/src/misc/iframe_component.ts

```diff
@@ -0,0 +1,81 @@
+import { I18nService } from '../abstractions/i18n.service';
+
+export abstract class IFrameComponent {
+    iframe: HTMLIFrameElement;
+    private connectorLink: HTMLAnchorElement;
+    private parseFunction = this.parseMessage.bind(this);
+
+    constructor(private win: Window, protected webVaultUrl: string, private path: string, private iframeId: string,
+        public successCallback?: (message: string) => any,
+        public errorCallback?: (message: string) => any, public infoCallback?: (message: string) => any) {
+        this.connectorLink = win.document.createElement('a');
+    }
+
+    stop() {
+        this.sendMessage('stop');
+    }
+
+    start() {
+        this.sendMessage('start');
+    }
+
+    sendMessage(message: any) {
+        if (!this.iframe || !this.iframe.src || !this.iframe.contentWindow) {
+            return;
+        }
+
+        this.iframe.contentWindow.postMessage(message, this.iframe.src);
+    }
+
+    base64Encode(str: string): string {
+        return btoa(encodeURIComponent(str).replace(/%([0-9A-F]{2})/g, (match, p1) => {
+            return String.fromCharCode(('0x' + p1) as any);
+        }));
+    }
+
+    cleanup() {
+        this.win.removeEventListener('message', this.parseFunction, false);
+    }
+
+    protected createParams(data: any) {
+        return new URLSearchParams({
+            data: this.base64Encode(JSON.stringify(data)),
+            parent: encodeURIComponent(this.win.document.location.href),
+            v: '1',
+        });
+    }
+
+    protected initComponent(params: URLSearchParams): void {
+        this.connectorLink.href = `${this.webVaultUrl}/${this.path}?${params}`;
+        this.iframe = this.win.document.getElementById(this.iframeId) as HTMLIFrameElement;
+        this.iframe.allow = 'publickey-credentials-get ' + new URL(this.webVaultUrl).origin;
```

Comment:
This is only needed for `WebAuthn`. While I don't see how this could be used negatively it's nice if we don't give more access than we absolutely need.